### PR TITLE
Add -[IGListAdapter objectForSectionController:] helper method -- iss…

### LINKED ISSUE #201: Add -[IGListAdapter objectForSectionController:] helper method
I've seen lots of instances where we want the object for a section controller. Typically you have to
```swift
let section = adapter.sectionFor(sectionController: controller)
let object = adapter.objectFor(section: section)
```
We can do this under the hood to streamline the API and give you:
```swift
let object = adapter.objectFor(sectionController: controller)
```

### PR DESCRIPTION
**Pull request by saraswatayu**
## Changes in this pull request

Add -[IGListAdapter objectForSectionController:] helper method

Fixes #201

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: Tests/IGListAdapterTests.m [hunk 1]
```diff
@@ -105,13 +105,27 @@ - (void)test_whenQueryingAdapter_withUnknownItem_thatSectionControllerIsNil {
...
 - (void)test_whenQueryingIndexPaths_withSectionController_thatPathsAreEqual {
     self.dataSource.objects = @[@0, @1, @2];
     [self.adapter performUpdatesAnimated:YES completion:nil];
     IGListSectionController <IGListSectionType> * second = [self.adapter sectionControllerForObject:@1];
-  NSArray *paths0 = [self.adapter indexPathsFromSectionController:second
-                                                       indexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(2, 4)]
-                                          adjustForUpdateBlock:NO];
+    NSArray *paths0 = [self.adapter indexPathsFromSectionController:second
+                                                            indexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(2, 4)]
+                                               adjustForUpdateBlock:NO];
```
Comment (rnystrom) [inline Tests/IGListAdapterTests.m]: Thanks!

File: Source/IGListAdapter.m [hunk 2]
```diff
@@ -308,6 +308,16 @@ - (NSUInteger)sectionForSectionController:(IGListSectionController <IGListSectio
     return [self.sectionMap sectionControllerForObject:object];
 }
 
+- (id)objectForSectionController:(IGListSectionController <IGListSectionType> *)sectionController {
+    IGAssertMainThread();
+    IGParameterAssert(sectionController != nil);
+    
+    NSUInteger section = [self.sectionMap sectionForSectionController:sectionController];
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: nit: `const`

File: Source/IGListAdapter.m [hunk 3]
```diff
@@ -308,6 +308,16 @@ - (NSUInteger)sectionForSectionController:(IGListSectionController <IGListSectio
...
+    IGAssert(section != NSNotFound, @"Section nil for section controller %@", sectionController);
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: Hmm if this assert will be too aggressive. Sometimes weird things happen and its probably ok to return `nil` in that case (for instance, the `-[IGListSectionMap objectForSection:]` method actually returns `nil` if the section is OOB).

I think we should probably omit this assert.

Comment (saraswatayu) [inline Source/IGListAdapter.m]: I based this off the implementation for `-[IGListAdapter objectAtSection:]`, which crashes if you pass in an index that's out of bounds. The assert I wrote actually wouldn't work, but there definitely needs to be a bounds check.

I'm assuming that's another issue that should be fixed?

Comment (rnystrom) [inline Source/IGListAdapter.m]: @saraswatayu that method should have been updated ([source here](https://github.com/Instagram/IGListKit/blob/master/Source/IGListAdapter.m#L307-L311)) so it doesn't assert.

Are you looking at the `stable` branch?

Comment (saraswatayu) [inline Source/IGListAdapter.m]: ahhh I see your commit on the SectionMap! I didn't realize my earlier rebase to my fork didn't quite work as expected; I'll update this real quick.

**Comment by jessesquires** [comment]
@saraswatayu -- Please add this to `CHANGELOG.md` too 😄

File: Source/IGListAdapter.h [hunk 4]
```diff
@@ -137,6 +137,15 @@ IGLK_SUBCLASSING_RESTRICTED
 - (__kindof IGListSectionController <IGListSectionType> * _Nullable)sectionControllerForObject:(id)object;
 
 /**
+ Fetch the object corresponding to a list in the feed. Constant time lookup.
```
Comment (jessesquires) [inline Source/IGListAdapter.h]: nit, let's say:

> `Returns the object corresponding to the specified section controller in the list. Constant time lookup.`

File: Source/IGListAdapter.h [hunk 5]
```diff
@@ -137,6 +137,15 @@ IGLK_SUBCLASSING_RESTRICTED
...
+ 
+ @param sectionController A list object.
```
Comment (jessesquires) [inline Source/IGListAdapter.h]: > `@param sectionController A section controller in the list.`

File: Source/IGListAdapter.h [hunk 6]
```diff
@@ -137,6 +137,15 @@ IGLK_SUBCLASSING_RESTRICTED
...
+ 
+ @return An object or nil.
```
Comment (jessesquires) [inline Source/IGListAdapter.h]: > `@return The object for the specified section controller, or `nil` if not found.`

**Comment by saraswatayu** [comment]
@jessesquires @rnystrom Thanks for your comments! It should now be updated - let me know if there's something else that catches your eye. 😄

**Pull request closed without merging**